x-pack/plugins/elastic_assistant/server/routes/post_actions_connector_execute.ts

```diff
@@ -19,128 +27,276 @@ import {
+                    getMessageFromRawResponse({
+                      rawContent: getMessageContentWithoutReplacements({
+                        messageContent: content,
+                        replacements: request.body.replacements as
+                          | Record<string, string>
+                          | undefined,
+                      }),
+                    }),
+                  ],
+                });
+              }
+            };
+          }
+
+          const connectorId = decodeURIComponent(request.params.connectorId);
+
+          // get the actions plugin start contract from the request context:
+          const actions = (await context.elasticAssistant).actions;
+
+          // if not langchain, call execute action directly and return the response:
+          if (
+            !request.body.isEnabledKnowledgeBase &&
+            !requestHasRequiredAnonymizationParams(request)
+          ) {
+            logger.debug('Executing via actions framework directly');
+            const result = await executeAction({
+              onMessageSent,
+              actions,
+              request,
+              connectorId,
+              params: {
+                subAction: request.body.params.subAction,
+                subActionParams: {
+                  ...request.body.params.subActionParams,
+                  messages: [
+                    ...(prevMessages ?? []),
+                    ...request.body.params.subActionParams.messages,
+                  ],
+                },
+              },
+            });
+            telemetry.reportEvent(INVOKE_ASSISTANT_SUCCESS_EVENT.eventType, {
+              isEnabledKnowledgeBase: request.body.isEnabledKnowledgeBase,
+              isEnabledRAGAlerts: request.body.isEnabledRAGAlerts,
+            });
+            return response.ok({
+              body: result,
+            });
+          }
+
+          // TODO: Add `traceId` to actions request when calling via langchain
+          logger.debug(
+            `Executing via langchain, isEnabledKnowledgeBase: ${request.body.isEnabledKnowledgeBase}, isEnabledRAGAlerts: ${request.body.isEnabledRAGAlerts}`
+          );
+
+          // Fetch any tools registered by the request's originating plugin
+          const pluginName = getPluginNameFromRequest({
+            request,
+            defaultPluginName: DEFAULT_PLUGIN_NAME,
+            logger,
+          });
+          const assistantTools = (await context.elasticAssistant).getRegisteredTools(pluginName);
+
+          // get a scoped esClient for assistant memory
+          const esClient = (await context.core).elasticsearch.client.asCurrentUser;
+
+          // convert the assistant messages to LangChain messages:
+          const langChainMessages = getLangChainMessages(
+            ([...(prevMessages ?? []), request.body.params.subActionParams.messages] ??
+              []) as unknown as Array<Pick<Message, 'content' | 'role'>>
+          );
```

Comment:
Why still grab the messages off the request vs query for them? 

From an API perspective, this allows you to 'in-line modify' the chat history to steer the output, but that all goes away once we move from the [BufferMemory](https://github.com/elastic/kibana/blob/fa47b572f3ecefc431d1327600c1e5dd804eafdf/x-pack/plugins/elastic_assistant/server/lib/langchain/execute_custom_llm_chain/index.ts#L54-L60) impl (which uses these raw messages) to a more robust conversation retrieval strategy (breaking down the prompt, fetching relevant messages/summary, etc).


This can be changed later as part of that effort, but just curious if there was any reason to keep this behavior for now. It is nice for testing still as you don't need to prime a persisted conversation and can just send it in the request, but again, we'll be re-visiting memory later. Which btw, would love to sync with you later on your thoughts around memory retrievers and memory in general! 🙂

Reply:
`request.body.params.subActionParams.messages` contains only the messages which was sent by user, while `prevMessages` contains queried persisted messages for the current conversation.
Could you please clarify what you mean by `query for them`?

> Which btw, would love to sync with you later on your thoughts around memory retrievers and memory in general!

I'll be more than happy do discuss the strategy.